src/main/java/com/google/devtools/build/lib/bazel/bzlmod/FactsAdapter.java

```diff
@@ -0,0 +1,71 @@
+// Copyright 2023 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package com.google.devtools.build.lib.bazel.bzlmod;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParser;
+import com.google.gson.TypeAdapter;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+import java.io.IOException;
+import net.starlark.java.eval.EvalException;
+import net.starlark.java.eval.Mutability;
+import net.starlark.java.eval.Starlark;
+import net.starlark.java.eval.StarlarkSemantics;
+import net.starlark.java.eval.StarlarkThread;
+import net.starlark.java.lib.json.Json;
+
+/** Helps serialize/deserialize {@link Facts}, which contains JSON-like Starlark values. */
+public class FactsAdapter extends TypeAdapter<Facts> {
+
+  private final Gson gson = new GsonBuilder().disableHtmlEscaping().create();
+
+  @Override
+  public void write(JsonWriter out, Facts facts) throws IOException {
+    String json;
+    try {
+      json = Json.INSTANCE.encode(facts.value());
```

Comment:
just curious -- how come we don't need the `INTERNAL_BAZEL_ONLY_UTF_8_BYTE_STRINGS` thing here?

Reply:
`encode` doesn't emit Unicode escape sequences for non-ASCII Unicode characters, it just dumps out their bytes verbatim - which happens to preserve raw UTF-8 bytes.